presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestMetadataQueryOptimizer.java

```diff
@@ -0,0 +1,282 @@
+                        ImmutableList.of(ImmutableList.of(p.symbol("key", BIGINT))),
+                        p.project(
+                                Assignments.of(
+                                        p.symbol("non_deterministic", BOOLEAN), p.expression("random()")),
+                                p.values(p.symbol("key", BIGINT)))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testDoesNotFireWhenNoTableScan()
+    {
+        tester.assertThat(metadataQueryOptimizer)
+                .on(p -> p.aggregate(
+                        ImmutableMap.of(p.symbol("COUNT", BIGINT),
+                                new Aggregation(COUNT_DISTINCT_FUNCTION_CALL, COUNT_SIGNATURE, Optional.empty())),
+                        ImmutableList.of(ImmutableList.of(p.symbol("key", BIGINT))),
+                        p.limit(5, p.values(p.symbol("key", BIGINT)))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testDoesNotFireWhenThereIsSort()
+    {
+        tester.assertThat(metadataQueryOptimizer)
+                .on(p -> p.aggregate(
+                        ImmutableMap.of(p.symbol("COUNT", BIGINT),
+                                new Aggregation(COUNT_DISTINCT_FUNCTION_CALL, COUNT_SIGNATURE, Optional.empty())),
+                        ImmutableList.of(ImmutableList.of(p.symbol("key", BIGINT))),
+                        p.filter(p.expression("key > 1"), p.tableScan(ImmutableList.of(p.symbol("key", BIGINT))))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void testMetadataQueryOptimization()
+    {
+        ConnectorId connectorId = tester.getCurrentConnectorId();
+        TpchTableHandle orders = new TpchTableHandle(connectorId.toString(), "orders", 1.0);
+        TableHandle ordersTable = new TableHandle(connectorId, orders);
+        TableLayoutHandle ordersTableLayout = new TableLayoutHandle(
+                tester.getCurrentConnectorId(),
+                TestingTransactionHandle.create(),
+                new TpchTableLayoutHandle(orders, Optional.empty()));
+
+        tester.assertThat(metadataQueryOptimizer)
+                .on(p -> p.aggregate(
+                        ImmutableMap.of(p.symbol("COUNT", BIGINT), new Aggregation(COUNT_DISTINCT_FUNCTION_CALL, COUNT_SIGNATURE, Optional.empty())),
+                        ImmutableList.of(ImmutableList.of()),
+                        p.tableScanWithTableLayout(
+                                ImmutableList.of(p.symbol("orderstatus", VARCHAR)),
+                                ImmutableMap.of(p.symbol("orderstatus", VARCHAR), ORDER_STATUS),
+                                ordersTable,
+                                ordersTableLayout)))
+                .matches(aggregation(ImmutableMap.of("COUNT", distinctFunctionCall("count", ImmutableList.of("orderstatus"))),
+                        values(ImmutableMap.of("orderstatus", 0))));
+    }
+
+    private class TestConnectorFactory
```

Comment:
I think the convention is classes named "TestXXX" are to test XXX.  and classes named "TestingXXX" is an XXX that is used for testing other things.